### PR TITLE
Handle batch documents updates

### PR DESCRIPTION
**Pull request by danipen**
Handle document changes that are between:

```chsarp
_textEditor.Document.BeginUpdate();
// stuff
_textEditor.Document.EndUpdate();
```

### CODE REVIEW

File: src/AvaloniaEdit.TextMate/TextEditorModel.cs [hunk 1]
```diff
@@ -124,22 +128,46 @@ private void DocumentOnChanged(object sender, DocumentChangeEventArgs e)
...
+            _invalidRange.SetInvalidRange(startLine, endLine);
+        }
+
+        void DocumentOnUpdateFinished(object sender, EventArgs e)
+        {
+            if (_invalidRange == null)
+                return;
+
+            InvalidateLineRange(_invalidRange.StartLine, _invalidRange.EndLine);
+            _invalidRange = null;
```
Comment (danipen) [inline src/AvaloniaEdit.TextMate/TextEditorModel.cs:167]: If `InvalidateLineRange` fails `_invalidRange` will never be set to null. Use try/finally:
```suggestion
            try
            {
                 InvalidateLineRange(_invalidRange.StartLine, _invalidRange.EndLine);
            }
            {
            finally          
            {           
                _invalidRange = null;
            }  
```

File: src/AvaloniaEdit.TextMate/TextEditorModel.cs [hunk 2]
```diff
@@ -152,5 +180,26 @@ void TokenizeViewPort()
                     _textView.VisualLines[_textView.VisualLines.Count - 1].LastDocumentLine.LineNumber - 1);
             }, DispatcherPriority.MinValue);
         }
+
+        internal class InvalidLineRange
+        {
+            internal int StartLine { get; set; }
+            internal int EndLine { get; set; }
```
Comment (danipen) [inline src/AvaloniaEdit.TextMate/TextEditorModel.cs:187]: You can use `private set;`

Comment (danipen) [inline src/AvaloniaEdit.TextMate/TextEditorModel.cs:187]: You can use `private set;`

File: test/AvaloniaEdit.Tests/TextMate/TextEditorModelTests.cs [hunk 3]
```diff
@@ -274,6 +280,78 @@ public void Replace_Document_Text_Should_Update_Line_Contents()
             Assert.AreEqual("two", textEditorModel.GetLineText(1));
             Assert.AreEqual("three", textEditorModel.GetLineText(2));
             Assert.AreEqual("four", textEditorModel.GetLineText(3));
+            textEditorModel.Dispose();
```
Comment (danipen) [inline test/AvaloniaEdit.Tests/TextMate/TextEditorModelTests.cs:283]: `textEditorModel.Dispose()` is missing for some tests.

Comment (danipen) [inline test/AvaloniaEdit.Tests/TextMate/TextEditorModelTests.cs:283]: `textEditorModel.Dispose()` is missing for some tests.

File: test/AvaloniaEdit.Tests/TextMate/TextEditorModelTests.cs [hunk 4]
```diff
@@ -274,6 +280,78 @@ public void Replace_Document_Text_Should_Update_Line_Contents()
...
+
+            TextEditorModel textEditorModel = new TextEditorModel(
+                textView, document, null);
+
+            document.Text = "puppy\npuppy\npuppy";
+
+            document.BeginUpdate();
+
+            document.Insert(0, "*");
+            Assert.AreEqual(0, textEditorModel.InvalidRange.StartLine);
```
Comment (danipen) [inline test/AvaloniaEdit.Tests/TextMate/TextEditorModelTests.cs:334]: Using multiple `Assert`, should include an assert message.

Comment (danipen) [inline test/AvaloniaEdit.Tests/TextMate/TextEditorModelTests.cs:334]: Using multiple `Assert`, should include an assert message.

File: test/AvaloniaEdit.Tests/TextMate/TextEditorModelTests.cs [hunk 5]
```diff
@@ -274,6 +280,78 @@ public void Replace_Document_Text_Should_Update_Line_Contents()
...
+            Assert.AreEqual(2, textEditorModel.InvalidRange.EndLine);
+
+            document.EndUpdate();
+            Assert.IsNull(textEditorModel.InvalidRange);
+
+            Assert.AreEqual("*puppy", textEditorModel.GetLineText(0));
+            Assert.AreEqual("*puppy", textEditorModel.GetLineText(1));
+            Assert.AreEqual("*puppy", textEditorModel.GetLineText(2));
+            textEditorModel.Dispose();
```
Comment (danipen) [inline test/AvaloniaEdit.Tests/TextMate/TextEditorModelTests.cs:317]: Use try/finally?

Comment (danipen) [inline test/AvaloniaEdit.Tests/TextMate/TextEditorModelTests.cs:317]: Use try/finally?

File: src/AvaloniaEdit.TextMate/TextEditorModel.cs [hunk 6]
```diff
@@ -124,22 +128,46 @@ private void DocumentOnChanged(object sender, DocumentChangeEventArgs e)
...
+            if (!_document.IsInUpdate)
+            {
+                InvalidateLineRange(startLine, endLine);
+                return;
+            }
+
+            // we're in a document change, store the max invalid range
+            if (_invalidRange == null)
+                _invalidRange = new InvalidLineRange(startLine, endLine);
+
```
Comment (danipen) [inline src/AvaloniaEdit.TextMate/TextEditorModel.cs:157]: You can return here.

Comment (danipen) [inline src/AvaloniaEdit.TextMate/TextEditorModel.cs:157]: You can return here.

**Review by danipen — COMMENTED** [review]
Requested changes

**Review by miryamGSM — APPROVED** [review]
LGTM

**Pull request merged**